File: pssm.py
# J Ludwig 2017
# My first python script

#/usr/bin/python

import sys #reading files
import re #matching the header line
import math #log
a=[] #contains mottif headers
b=[] #contains mottif sequences
header_counter=-1
with open(sys.argv[1], 'r') as mottif_file: #readfile from input argument 1 as mottif data
	for line in mottif_file.readlines():
		line = line.rstrip('\n') #remove "\n" 
		if re.match('>',line): #detecting headers from fasta file
			a.append(line) #adding mottif header line data
			flag=0 #flag used to identify the next line of data as sequence data
			header_counter = header_counter + 1 #counting the number of headers in mottif file
		elif flag==0:
			b.append(line) #adding sequence data
			flag=1 #flag to continue adding sequence data until next ">"
		else:
			b[header_counter] = b[header_counter] + line #adding sequence data; this method was used over append as append added "," between entries
maxb = len(max(b)) #variable used later in loop; set to the max length of the longest detected mottif sequence
c=[] #contains sequence headers
d=[] #contains sequence sequences/data
Gcount=0 #variable containing count of G nucleotide
Ccount=0 #variable containing count of C nucleotide
Acount=0 #variable containing count of A nucleotide
Tcount=0 #variable containing count of T nucleotide 
Totalcount=0 #variable containing count of all nucleotide
header_counter=-1
with open(sys.argv[2], 'r') as sequence_file: #readfile from input argument 2 as sequence data
	for line in sequence_file.readlines():
		line=line.rstrip() #remove "\n" 
		if re.match('>',line): #detecting headers from fasta file
			c.append(line) #adding sequence header line
			flag=0 #flag used to identify the next line of data as sequence data
			header_counter = header_counter + 1 #counting the number of headers in sequence file
		elif flag==0:
			d.append(line) #adding sequence data
			flag=1
			Gcount = Gcount + (line.count('G')) #counting G nucleotide
			Ccount = Ccount + (line.count('C')) #counting C nucleotide
			Acount = Acount + (line.count('A')) #counting A nucleotide
			Tcount = Tcount + (line.count('T')) #counting T nucleotide
			Totalcount = Totalcount + (line.count('')) #counting Total nucleotide
		else:
			d[header_counter] = d[header_counter] + line #adding sequence data; this method was used over append as append added "," between entries
			Gcount = Gcount + (line.count('G')) #counting G nucleotide
			Ccount = Ccount + (line.count('C')) #counting C nucleotide
			Acount = Acount + (line.count('A')) #counting A nucleotide
			Tcount = Tcount + (line.count('T')) #counting T nucleotide
			Totalcount = Totalcount + (line.count('')) #counting Total nucleotide
if re.match('\d+%',sys.argv[3]):
	prescore=sys.argv[3]
	score=prescore.rstrip(("%"))
else:
	score = int(sys.argv[3]) #score from input argument 3
maxd = len(max(d)) #variable used later in loop; set to the max length of the longest detected mottif sequence
Gpercent = Gcount/Totalcount #calculate % of Genome consisting of G
Cpercent = Ccount/Totalcount #calculate % of Genome consisting of C
Apercent = Acount/Totalcount #calculate % of Genome consisting of A
Tpercent = Tcount/Totalcount #calculate % of Genome consisting of T
x=0 #variable for next loop
y=0 #variable for next loop
Matrix = [[0.00 for i in range(4)] for i in range(maxb)] #create array with 4 columns and number of rows = longest mottif sequence length; 0.00 is value for all entry
while (x< maxb):
	Matrix[y][0] = Apercent #set value of first column to background % of A in test sequence
	Matrix[y][1] = Cpercent #set value of first column to background % of C in test sequence
	Matrix[y][2] = Gpercent #set value of first column to background % of G in test sequence
	Matrix[y][3] = Tpercent #set value of first column to background % of T in test sequence
	x = x + 1
	y = y + 1 
x=0 #variable for next loop
while (x < len(b)): #note currently only lower case
	y = 0 #variable for loop
	for sequence in b[x]: #note these are currently set to lower case
		if sequence == 'a':
			Matrix[y][0] = Matrix[y][0] + 1 #counting number of A nucleotide in motiff sequence
		if sequence == 'c':
			Matrix[y][1] = Matrix[y][1] + 1 #counting number of C nucleotide in motiff sequence
		if sequence == 'g':
			Matrix[y][2] = Matrix[y][2] + 1 #counting number of G nucleotide in motiff sequence
		if sequence == 't':
			Matrix[y][3] = Matrix[y][3] + 1 #counting number of T nucleotide in motiff sequence
		y = y + 1
	x = x + 1
i = 0
print ("@@Count of nucleotides by position in motiff file")
print ("{:<7}{:<7}{:<7}{:<7}{:<7}".format("POS","A","C","G","T"))
while (i < maxb):	
	print ("{:3}{!s:>7.5}{!s:>7.5}{!s:>7.5}{!s:>7.5}".format(i+1,Matrix[i][0],Matrix[i][1],Matrix[i][2],Matrix[i][3])) #print count of nucleotides in mottif
	i = i + 1
f = 0
Matrix2 = [[0 for i in range(4)] for i in range(maxb)] 
while (f < maxb): #calculate frequency of test sequence nucleotides
	Matrix2[f][0] = Matrix[f][0]/(Matrix[f][0]+Matrix[f][1]+Matrix[f][2]+Matrix[f][3]) 
	Matrix2[f][1] = Matrix[f][1]/(Matrix[f][0]+Matrix[f][1]+Matrix[f][2]+Matrix[f][3]) 
	Matrix2[f][2] = Matrix[f][2]/(Matrix[f][0]+Matrix[f][1]+Matrix[f][2]+Matrix[f][3]) 
	Matrix2[f][3] = Matrix[f][3]/(Matrix[f][0]+Matrix[f][1]+Matrix[f][2]+Matrix[f][3]) 
	f = f + 1
f = maxb - 1
g=0
Matrix3 = [[0 for i in range(4)] for i in range(maxb)] 
while (f > -1): #invert instead of reverse compliment the test sequence to test for both directions
	Matrix3[f][0] = Matrix[g][0]/(Matrix[g][0]+Matrix[g][1]+Matrix[g][2]+Matrix[g][3]) 
	Matrix3[f][1] = Matrix[g][1]/(Matrix[g][0]+Matrix[g][1]+Matrix[g][2]+Matrix[g][3]) 
	Matrix3[f][2] = Matrix[g][2]/(Matrix[g][0]+Matrix[g][1]+Matrix[g][2]+Matrix[g][3]) 
	Matrix3[f][3] = Matrix[g][3]/(Matrix[g][0]+Matrix[g][1]+Matrix[g][2]+Matrix[g][3]) 
	g = g + 1
	f = f - 1
	
print ("@@ Frequency Matrix")
print ("{:<7}{:<7}{:<7}{:<7}{:<7}".format("POS","A","C","G","T"))
h = 0
while (h < maxb):	
	print ("{:3}{!s:>7.5}{!s:>7.5}{!s:>7.5}{!s:>7.5}".format(h+1,Matrix2[h][0],Matrix2[h][1],Matrix2[h][2],Matrix2[h][3])) #print frequency of nucleotide of test sequence

	h = h + 1
z=0
print ("@@PSSM")
print ("{:<7}{:<7}{:<7}{:<7}{:<7}".format("POS","A","C","G","T"))
while (z < maxb): #print PSSM
	print ("{:3}{!s:>7.5}{!s:>7.5}{!s:>7.5}{!s:>7.5}".format(z + 1,(math.log2((Matrix2[z][0]) / Apercent)),(math.log2((Matrix2[z][1]) / Cpercent)),(math.log2((Matrix2[z][2]) / Gpercent)),(math.log2((Matrix2[z][3]) / Tpercent))))
	z = z+1


start=0
finish=maxb #sliding window set to length of longest mottif sequence
while (finish<maxd): 
	m=0
	thing = (d[0][start:finish]) #get the nucleotides from the test sequence between start position and end position
	x=0
	x2 = 0 
	for thing2 in thing:
		if thing2 == 'A':
			s =(math.log2((Matrix2[m][0]) / Apercent))
			s2 = (math.log2((Matrix3[m][3]) / Tpercent))
		if thing2 == 'C':
			s =(math.log2((Matrix2[m][1]) / Cpercent))
			s2 = (math.log2((Matrix3[m][2]) / Gpercent))
		if thing2 == 'G':
			s =(math.log2((Matrix2[m][2]) / Gpercent))
			s2=(math.log2((Matrix3[m][1]) / Cpercent))
		if thing2 == 'T':
			s =(math.log2((Matrix2[m][3]) / Tpercent))
			s2=(math.log2((Matrix3[m][0]) / Apercent))
		m = m + 1
		x = x + s
		x2 = x2 + s2
	
	if x > score: #{:18} 18 is length to display, {:^5d} center align 5 digits
		print ("{:18}{:^8d}{:^11}{:8d}{:^7}{:5.1f}{:^10}{:}".format("Match starting at",start+1, "Ending at", finish, "Score", x, "Sequence", d[0][start:finish]))
	if x2 > score:
		print ("{:18}{:^8d}{:^11}{:8d}{:^7}{:5.1f}{:^10}{:}".format("Match starting at",start+1, "Ending at", finish, "Score", x2, "Sequence", d[0][start:finish]))
	start = start + 1
	finish = finish + 1

File: PSSM.cpp
// J Ludwig 2018

#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <math.h>  // log
#include <iomanip>   // for setprecision

using namespace std;
void printer(int start, int finish, string sample_data[]);
void printer2(int i, int max_seq_length, string* motif_data[]);

int main(int argc, char* argv[])
{

    cout << setprecision(3);  //number of digits to print in output
    cout << showpoint;
    cout << fixed;


    double score = stod((argv[3]));

    ifstream file(argv[1]);                              //Read infile and get size of file
    //ifstream file ("ecoK12-MG1655.fasta");

    string sample_name;
    getline(file, sample_name, '\n');
    string temp;
    char c;
    int temp_char_counter{0};
    while(getline(file, temp, '\n')){
        stringstream ss (temp);
        while(ss >> c){
            ++temp_char_counter;
        }
    }

    file.close();
    long int Acount{0};
    long int Tcount{0};
    long int Gcount{0};
    long int Ccount{0};
    double Totalcount{0};
    ifstream file2(argv[1]);                       //Read query infile and store data
    //ifstream file2 ("ecoK12-MG1655.fasta");
    string* sample_data = new string[temp_char_counter];
    string temp2;
    int temp_counter{0};
    getline(file2, temp2);
    while(getline(file2, temp2, '\n')){
        stringstream ss (temp2);
        while(ss >> c){
            sample_data[temp_counter] = c;
            ++temp_counter;
            ++Totalcount;
            if(c == 'A' || c == 'a'){
                ++Acount;
            } else if(c == 'T' || c == 't'){
                ++Tcount;
            } else if(c == 'G' || c == 'g'){
                ++Gcount;
            } else if(c == 'C' || c == 'c'){
                ++Ccount;
            }
        }
    }
    file2.close();
    double Apercent = Acount / Totalcount;
    double Tpercent = Tcount / Totalcount;
    double Gpercent = Gcount / Totalcount;
    double Cpercent = Ccount / Totalcount;

    //cout << "A count " << Acount << "\n";
    //cout << "Tot count " << Totalcount << "\n";
    //cout << "A% " << Apercent << "\n";

    string temp3;
    ifstream file3(argv[2]);                           //Read motif (database) infile and get size. Can read both fasta and txt files
    //ifstream file3 ("sigma54.fasta");
    getline(file3, temp3, '\n');
    stringstream ss (temp3);
    ss >> c;
    int row_counter{0};
    int max_seq_length{0};
    bool flag_fasta = 0;
    if(c == '>'){
        flag_fasta = 1;
        while(getline(file3, temp3, '\n')){
            int temp_seq_length{0};
            stringstream ss (temp3);
            while(ss >> c){
                ++temp_seq_length;
            }
            getline(file3, temp, '\n');
            ++row_counter;
            if(temp_seq_length > max_seq_length){
                max_seq_length = temp_seq_length;
            }
        }
    } else {
        ++row_counter;
        int temp_seq_length{0};
        while(ss >> c){
            ++temp_seq_length;
        }
        if(temp_seq_length > max_seq_length){
            max_seq_length = temp_seq_length;
        }
        while(getline(file3, temp3, '\n')){
            int temp_seq_length{0};
            stringstream ss (temp3);
            while(ss >> c){
                ++temp_seq_length;
            }
            ++row_counter;
            if(temp_seq_length > max_seq_length){
                max_seq_length = temp_seq_length;
            }
        }
    }

    //cout << "row counter " << row_counter << "\n";
    //cout << "max seq length " << max_seq_length << "\n";

    file3.close();

    ifstream file4(argv[2]);                                //Read motif infile and store data
    //ifstream file4 ("sigma54.fasta");
    string* motif_name = new string[row_counter];
    string** motif_data = new string*[row_counter];
    for(int i = 0; i < row_counter; ++i){
        motif_data[i] = new string[max_seq_length];
    }
    double** freq_matrix = new double*[max_seq_length];
    for(int i = 0; i < max_seq_length; ++i){
        freq_matrix[i] = new double[4]{Apercent, Tpercent, Gpercent, Cpercent};
    }
    temp_counter = 0;
    if(flag_fasta == 1){                                                                           //Two similar loops. One for fasta format and one for other format
        bool alt = 1;
        while(getline(file4, temp3, '\n')){
            if(alt == 1){
                motif_name[temp_counter] = temp3;
                alt = 0;
            } else {
                int temp_counter2{0};
                stringstream ss (temp3);
                while(ss >> c){
                    motif_data[temp_counter][temp_counter2] = c;                                    //Count number of each nucleotide
                    if(c == 'A' || c == 'a'){
                        freq_matrix[temp_counter2][0] = freq_matrix[temp_counter2][0] + 1;
                    } else if(c == 'T' || c == 't'){
                        freq_matrix[temp_counter2][1] = freq_matrix[temp_counter2][1] + 1;
                    } else if(c == 'G' || c == 'g'){
                        freq_matrix[temp_counter2][2] = freq_matrix[temp_counter2][2] + 1;
                    } else if (c == 'C' || c == 'c'){
                        freq_matrix[temp_counter2][3] = freq_matrix[temp_counter2][3] + 1;
                    }
                    ++temp_counter2;
                }
                alt = 1;
                ++temp_counter;
            }
        }
    } else {
        while(getline(file4, temp3, '\n')){
            int temp_counter2{0};
            stringstream ss (temp3);
            while(ss >> c){
                motif_data[temp_counter][temp_counter2] = c;
                if(c == 'A' || c == 'a'){
                    freq_matrix[temp_counter2][0] = freq_matrix[temp_counter2][0] + 1;
                } else if(c == 'T' || c == 't'){
                    freq_matrix[temp_counter2][1] = freq_matrix[temp_counter2][1] + 1;
                } else if(c == 'G' || c == 'g'){
                    freq_matrix[temp_counter2][2] = freq_matrix[temp_counter2][2] + 1;
                } else if (c == 'C' || c == 'c'){
                    freq_matrix[temp_counter2][3] = freq_matrix[temp_counter2][3] + 1;
                }
                ++temp_counter2;
            }
        ++temp_counter;
        }
    }

    /* Uncomment to see motif data printed to stdout
    for(int i = 0; i < row_counter; ++i){
        cout << motif_name[i] << "\n";
        for(int j = 0; j < max_seq_length; ++j){
            cout << motif_data[i][j];
        }
        cout << "\n";
    }
    */

    cout << "\n";                                                           //Print frequency matrix
    cout << "Freq matrix" << "\n";
    cout << "\t\t" << "A \t" << "T \t" << "G \t" << "C \t" << "\n";
    for(int i = 0; i < max_seq_length; ++i){
        cout << "Position " << i << "\t";
        for(int j = 0; j < 4; ++j){
            cout << setw(7) << freq_matrix[i][j] << "\t";
        }
        cout << "\n";
    }
    cout << "\n";


    //cout << "Sample A count " << Acount << "\n";
    //cout << "Sample T count " << Tcount << "\n";
    //cout << "Sample G count " << Gcount << "\n";
    //cout << "Sample C count " << Ccount << "\n";
    //cout << "Sample Total count " << Totalcount << "\n";

    for(int i = 0; i < max_seq_length; ++i){
        double row_sum{0};
        for(int j = 0; j < 4; ++j){
            row_sum = row_sum + freq_matrix[i][j];
        }
        for(int k = 0; k < 4; ++k){
            freq_matrix[i][k] = freq_matrix[i][k] / row_sum;
        }
    }

    //for(int i = 0; i < max_seq_length; ++i){
    //    cout << "\n";
    //    for(int j = 0; j < 4; ++j){
    //        cout << freq_matrix[i][j];
    //    }
    //}

    double** freq_matrix_inverse = new double*[max_seq_length];
    for(int i = 0; i < max_seq_length; ++i){
        freq_matrix_inverse[i] = new double[4];
    }

    int g = 0;
    int f = max_seq_length-1;                                                //Transpose the matrix instead of reverse comp
    while(f > -1){
        double background = freq_matrix[g][0] + freq_matrix[g][1] + freq_matrix[g][2] + freq_matrix[g][3];
        freq_matrix_inverse[f][0] = freq_matrix[g][0] / background;
        freq_matrix_inverse[f][1] = freq_matrix[g][1] / background;
        freq_matrix_inverse[f][2] = freq_matrix[g][2] / background;
        freq_matrix_inverse[f][3] = freq_matrix[g][3] / background;
    }

    cout << "PSSM matrix" << "\n";                                               //Print PSSM
    cout << "\t\t" << "A \t" << "T \t" << "G \t" << "C \t" << "\n";
    for(int i = 0; i < max_seq_length; ++i){
        cout << "Position " << i << "\t";
        cout << setw(7) << log2(freq_matrix[i][0] / Apercent) << "\t";
        cout << setw(7) << log2(freq_matrix[i][1] / Tpercent) << "\t";
        cout << setw(7) << log2(freq_matrix[i][2] / Gpercent) << "\t";
        cout << setw(7) << log2(freq_matrix[i][3] / Cpercent) << "\n";
    }
    cout << "\n";

    for(int i = 0; i < row_counter; ++i){                                   //Print motif score
        int m{0};
        double x{0};
        while(m < max_seq_length){
            string thing = motif_data[i][m];
            if(thing == "A" || thing == "a"){
                x  = x  + log2(freq_matrix[m][0] / Apercent);
            } else if(thing == "T" || thing == "t"){
                x  = x  + log2(freq_matrix[m][1] / Tpercent);
            } else if(thing == "G" || thing == "g"){
                x  = x  + log2(freq_matrix[m][2] / Gpercent);
            } else if(thing == "C" || thing == "c"){
                x  = x  + log2(freq_matrix[m][3] / Cpercent);
            }
            ++m;
        }
        if(x > score){
            cout << "Motif score " << x << " ";

            printer2(i, max_seq_length, motif_data);
            cout << "\n";
        }
    }

    cout << "\n";
    int start = 0;
    int finish = max_seq_length;
    
}

